### DISCUSSION TITLE
Relax restrictions to allow T* extension methods to be defined from C#

### DISCUSSION THREAD

**Discussion by DaZombieKiller** [dotnet/csharplang#5367] [Language Ideas]
C# currently supports consuming `T*` extension methods if they are defined in IL, but attempting to define them in C# will result in a compiler error:
```cs
public static unsafe class PointerExtensions
{
    // CS1103: The first parameter of an extension method cannot be of type 'int*'
    public static int* Nop(this int* self) => self;

    // CS1103: The first parameter of an extension method cannot be of type 'T*'
    public static T* Nop<T>(this T* self) where T : unmanaged => self;
}
```
If you define these in IL, they can be used as expected:
```cs
int* pointer = /* ... */;
pointer.Nop();
byte* bytePointer = /* ... */;
bytePointer.Nop();
```
I propose relaxing the restrictions to allow them to be defined in C#.

**Comment by siegfriedpammer** [comment]
Note that your example is slightly incorrect. Calling methods on pointers either requires `->` instead of `.` (as in `bytePointer->Nop();`, or you'd have to dereference the invocation target first, e.g., `(*bytePointer).Nop();`

What you propose is already possible today using `this ref`. See https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwAIBMAjALABQeAzAASHUDC1A3udW7TQK4wDOEAZgFNaAFmoBZABSwE1AFQAPAJTNW7dQoC0APgByAewAOkpQG416tpMVKAdAeNmL1AL7k3ZclVpEAbLQJqAFEFBEFeKH1eVTJLbzw/UWoHSQQACygeagAnQX5qGWplZg8XIA=

Hope this helps!

**Reply by DaZombieKiller** [reply]
My example is correct for a pointer extension method (it is the syntax that you can use today if you define a pointer extension method in IL.)
`ref this` extension methods aren't quite the same thing and do not guarantee that the value is `fixed`, so they aren't suitable for many of the cases you would need to use pointers for.

**Reply by AraHaan** [reply]
I agree, unless we could do this to ensure that they are fixed if you make the first parameter ``fixed ref this`` to denote that "This can only be used when the pointer is fixed / pinned".

**Comment by Atomiz2002** [comment]
In the admittedly rare occasions this would be used (I'm here because of one), I also find it quite annoying that it's theoretically possible but practically not.

**Reply by lgztx96** [reply]
In C# 14, it works.
![image](https://github.com/user-attachments/assets/c23df2ec-a617-4551-89f7-318ce009eeb0)
![image](https://github.com/user-attachments/assets/c6ce426f-4d7c-4b90-9802-2c455abe5e8e)

**Reply by DaZombieKiller** [reply]
@lgztx96 It was not intended for this to be allowed, and as far as I know it no longer works after .NET 10 preview 3.

**Reply by lgztx96** [reply]
> 它不是为了允许这样做，据我所知，它在 .NET 10 预览版 3 之后不再有效。

That's really bad news.

**Comment by DanielShuey** [comment]
At the moment I am substituting with `this ref`. It will automatically typecast, but I'm not sure how this affects performance.

If not the only issue is that the code looks inconsistent. This is a problem because I am treating all my structs like pointers, but when I read it in an extension method it looks like some are passed by value.

I am transitioning into only using pointers and never using ref. So this would be useful for me.
